### DISCUSSION TITLE
Practice - Create a class diagram for BookYourShow

### DISCUSSION THREAD

**Discussion by kanmaytacker** [kanmaytacker/design-questions#8] [General]
What?
* Identify the major entities and their attributes in a ticket booking management system ([BookMyShow](https://in.bookmyshow.com/))
* Create a class diagram for the system.

Why?
Interesting design and a popular question in interviews.

How?
Follow the instructions [here](https://github.com/kanmaytacker/design-questions/blob/master/notes/06-design-book-my-show-worksheet.md#class-diagrams).

**Comment by Faizahmadfaiz** [comment]
**Class Names and it's attributes:**

### Entities and Attributes

1. City
    - name
    - state
2. Cinema
    - name
    - type
    - City
3. Hall
    - name
    - type
    - Cinema
4. Movie
    - name
    - cast
    - type
    - duration
    - releaseDate
    - language
    - Genre

5. Seat
    - type
    - status
    - Hall
6. User
    - name
    - age
    - Booking
7. Show
    - time
    - Hall
8. Ticket
    - Booking
9. Payment
    - method
    - time
    - Booking
10. Coupon
    - code
11. Booking
    - time
    - price
12. Genre
    - name

**Comment by navnnkk** [comment]
```mermaid
classDiagram
    class City {
        -String name
        -String code
        -String is_active
        +getCity() String
    }

    class Theater {
        -String name
        -Integer city_id
    }

    class Movie {
        -String name
        -Integer rating
        -String genre
        -String plot
        -String duration
        -String language
    }

    class Hall {
        -Integer hall_number
        -Integer theater_id
        -Integer movie
    }

    class Seat {
        -enum seat_type
        -Integer hall_id
        -String seat_number
    }

    class Shows {
        -Integer movie_id
    }

    class Ticket {
        -Integer ticket_number
        -Integer amount
        -Integer show_id
    }

    class User {
        -first_name
        -last_name
        -email_id
    }

    class Coupan {
        String Code
        bool is_active
    }

```

**Comment by kanmaytacker** [comment]
Next steps -

1. [Optional] Convert entity list to a mermaid class diagram
2. [Optional] Or add the relationships to the existing class diagram [here](https://github.com/kanmaytacker/design-questions/blob/master/notes/06-design-book-my-show.md#class-diagrams).
3. [Useful] Create the classes from the diagram in Java etc.

[Current worksheet](https://github.com/kanmaytacker/design-questions/blob/master/notes/06-design-book-my-show-worksheet.md#class-diagrams)

**Comment by vkartik** [comment]
![image](https://user-images.githubusercontent.com/11883108/200000115-a71aa65b-6a44-4625-a2ef-76150027bd14.png)

**Comment by Mariselvam** [comment]
@kanmaytacker   I have just added relationship between the classes , could you check if this is looking good..?

![bms-classDiag-relationships](https://user-images.githubusercontent.com/1678425/201516211-0080b1aa-be13-41e7-96bd-aee788012580.png)

**Comment by Mariselvam** [comment]
@kanmaytacker   couple of questions in our class diagram for BMS.

1) Why the Movie class is having 'Shows' as its attribute ..?  to support search functionality ..?

2) Ticket class is having reference to 'Payment' class and 'Payment' class is having reference to 'Ticket' class, why this is so ..? isn't creating cyclic dependency  at the DB entities level ..?

3) Are we missing 'search' functionality in our class design ...?
